core/kernel/transfer_list.c

```diff
@@ -0,0 +1,542 @@
+	tl->alignment = TRANSFER_LIST_INIT_MAX_ALIGN; /* initial max align */
+	tl->size = sizeof(*tl); /* initial size is the size of header */
+	tl->max_size = max_size;
+
+	transfer_list_update_checksum(tl);
+
+	return tl;
+}
+
+/*******************************************************************************
+ * Relocating a transfer list to a reserved memory region specified
+ * Compliant to 2.4.6 of Firmware handoff specification (v0.9)
+ * Return pointer to the relocated transfer list or NULL on error
+ ******************************************************************************/
+struct transfer_list_header *
+transfer_list_relocate(struct transfer_list_header *tl, paddr_t pa,
+		       size_t max_size)
+{
+	paddr_t new_addr = 0;
+	struct transfer_list_header *new_tl = NULL;
+	size_t new_max_size = 0;
+
+	if (!tl || !pa || !max_size)
+		return NULL;
+
+	new_addr = get_align_base_addr(pa, tl);
+	new_max_size = max_size - (new_addr - pa);
+
+	/* The new space is not sufficient for the TL */
+	if (tl->size > new_max_size)
+		return NULL;
+
+	new_tl = core_mmu_add_mapping(MEM_AREA_TRANSFER_LIST, new_addr,
+				      new_max_size);
+	if (!new_tl)
+		return NULL;
+
+	memmove(new_tl, tl, tl->size);
+	new_tl->max_size = new_max_size;
+
+	transfer_list_update_checksum(new_tl);
+	transfer_list_unmap_sync(tl);
+
+	return new_tl;
+}
+
+/*******************************************************************************
+ * Verifying the header of a transfer list
+ * Compliant to 2.4.1 of Firmware handoff specification (v0.9)
+ * Return transfer list operation status code
+ ******************************************************************************/
+int transfer_list_check_header(const struct transfer_list_header *tl)
+{
+	if (!tl)
+		return TL_OPS_NON;
+
+	if (tl->signature != TRANSFER_LIST_SIGNATURE) {
+		EMSG("Bad transfer list signature %#"PRIx32, tl->signature);
+		return TL_OPS_NON;
+	}
+
+	if (!tl->max_size) {
+		EMSG("Bad transfer list max size %#"PRIx32, tl->max_size);
+		return TL_OPS_NON;
+	}
+
+	if (tl->size > tl->max_size) {
+		EMSG("Bad transfer list size %#"PRIx32, tl->size);
+		return TL_OPS_NON;
+	}
+
+	if (tl->hdr_size != sizeof(struct transfer_list_header)) {
+		EMSG("Bad transfer list header size %#"PRIx8, tl->hdr_size);
+		return TL_OPS_NON;
+	}
+
+	if (!transfer_list_verify_checksum(tl)) {
+		EMSG("Bad transfer list checksum %#"PRIx8, tl->checksum);
+		return TL_OPS_NON;
+	}
+
+	if (tl->version == 0) {
+		EMSG("Transfer list version is invalid");
+		return TL_OPS_NON;
+	} else if (tl->version == TRANSFER_LIST_VERSION) {
+		DMSG("Transfer list version is valid for all operations");
+		return TL_OPS_ALL;
+	} else if (tl->version > TRANSFER_LIST_VERSION) {
+		DMSG("Transfer list version is valid for read-only");
+		return TL_OPS_RO;
+	}
+
+	DMSG("Old transfer list version is detected");
+	return TL_OPS_CUS;
+}
+
+/*******************************************************************************
+ * Enumerate the next transfer entry
```

Comment:
It would be nice that argument `last` is also described. I guess here last refers to the last parsed TE, from where we want to find the next TE.

Reply:
I realize that 'last' might be not very accurate, I should use 'current' instead - We want to pass in the 'current' TE and then find its 'next'.
I will update it with the comment.